src/main/java/com/conveyal/r5/analyst/Grid.java

```diff
@@ -335,4 +359,139 @@ public static Polygon getPixelGeometry (int x, int y, int zoom) {
+        if (!Stream.of(headers).filter(h -> h.equals(latField)).findAny().isPresent()) {
+            LOG.info("Lat field not found!");
+            halt(400, "Lat field not found");
+        }
+
+        if (!Stream.of(headers).filter(h -> h.equals(lonField)).findAny().isPresent()) {
+            LOG.info("Lon field not found!");
+            halt(400, "Lon field not found");
+        }
+
+        Envelope envelope = new Envelope();
+
+        // Keep track of which fields contain numeric values
+        Set<String> numericColumns = Stream.of(headers).collect(Collectors.toCollection(HashSet::new));
+        numericColumns.remove(latField);
+        numericColumns.remove(lonField);
+
+        int i = 0;
+        while (reader.readRecord()) {
+            if (++i % 10000 == 0) LOG.info("{} records", human(i));
+
+            envelope.expandToInclude(parseDouble(reader.get(lonField)), parseDouble(reader.get(latField)));
+
+            for (Iterator<String> it = numericColumns.iterator(); it.hasNext();) {
+                String field = it.next();
+                String value = reader.get(field);
+                if (value == null || "".equals(value)) continue; // allow missing data
+                try {
+                    // TODO also exclude columns containing negatives?
+                    parseDouble(value);
+                } catch (NumberFormatException e) {
+                    it.remove();
+                }
+            }
+        }
+
+        reader.close();
+
+        // We now have an envelope and know which columns are numeric
+        // Make a grid for each numeric column
+        Map<String, Grid> grids = numericColumns.stream()
+                .collect(
+                        Collectors.toMap(
+                                c -> c,
+                                c -> new Grid(
+                                        zoom,
+                                        envelope.getMaxY(),
+                                        envelope.getMaxX(),
+                                        envelope.getMinY(),
+                                        envelope.getMinX()
+                                )));
+
+        // read it again, Sam - reread the CSV to get the actual values and populate the grids
+        reader = new CsvReader(new BufferedInputStream(new FileInputStream(csvFile)), Charset.forName("UTF-8"));
+        reader.readHeaders();
+
+        i = 0;
+        while (reader.readRecord()) {
+            if (++i % 10000 == 0) LOG.info("{} records", human(i));
+
+            double lat = parseDouble(reader.get(latField));
+            double lon = parseDouble(reader.get(lonField));
+
+            for (String field : numericColumns) {
+                String value = reader.get(field);
+
+                double val;
+
+                if (value == null || "".equals(value)) {
+                    val = 0;
+                } else {
+                    val = parseDouble(value);
```

Comment:
If some column contains the occasional non-numeric junk, that columns will be deemed non-numeric, and won't be read in this second pass (though that prevents things from blowing up here). Maybe we should detect that case and throw a descriptive error that can reach the end user.